---
title: "Using recoup with continuous regions (ChIP-Seq)"
author: "Panagiotis Moulos"
date: "`r BiocStyle::doc_date()`"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Using recoup with continuous regions (ChIP-Seq)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

Construction of genomic coverage profiles in continuous regions
===============================================================

This tutotial demonstrates a few examples of how you can use `recoup` to create
average genomic cpverage profiles and heatmaps from short read sequences for 
genomic regions who are continuous. Such regions are ChIP-Seq peak areas, DNA
methylation areas, PolII profiles over gene bodies or other custom regions which
are __continuous__ (i.e. not spliced like intron-exon areas). Although most 
parameters are self-explanarory, this tutorial is mean to be done along with 
reading the man page of the `recoup` function.

## Get some data

The small dataset attached to the package is not realistic and is there only to
pass the tests and comply with Bioconductor guidelines. Creating genomic 
coverage profiles generally takes some time, but not something extreme. All the 
process can be significantly get faster of you work or have access to a system 
with multiple cores. We are working to make things faster! Any suggestions are 
most welcome!

For the rest of this tutorial, it is supposed that you have uncompressed the tar
archive with example data described in the introductory page, in a directory 
called `/home/me/recoup_tutorial`. We will use the H4K20me1 data located in `/home/me/recoup_tutorial/chipseq`. Have a look at the
[introductory tutorial page](A_recoup_intro.html) for a short description of the
test data.

## Running recoup with ChIP-Seq data

The `recoup` function can be run either using a text __tab__ delimited file (no
other delimiters supported at this point) or a list of lists, where each sublist
has some fields (mandatory and optional) with sample data. Have a look at the 
`recoup` man page for the format of the file and the list. A configuration file
 example can be found in the downloaded test data archive with the name 
 `test_config.txt`. It has the following format:

| id              | name            | file                                                  | format | color     |
|-----------------|-----------------|-------------------------------------------------------|--------|-----------|
| WT_H4K20me1     | WT H4K20me1     | /home/me/recoup_tutorial/chipseq/WT_H4K20me1.bam     | bam    | "#EE0000" |
| Set8KO_H4K20me1 | Set8KO H4K20me1 | /home/me/recoup_tutorial/chipseq/Set8KO_H4K20me1.bam | bam    | "#00BB00" |

In case of a list, it has the same field as the configuration file. An input 
list example:

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
library(recoup)

test.path <- "/home/me/recoup_tutorial/chipseq"

chip.input <- list(
    list(
        id="WT_H4K20me1",
        name="WT H4K20me1",
        file=file.path(test.path,"WT_H4K20me1.bam"),
        format="bam"
    ),
    list(
        id="Set8KO_H4K20me1",
        name="Set8KO H4K20me1",
        file=file.path(test.path,"Set8KO_H4K20me1.bam"),
        format="bam"
    )
)
```

With this input, default ggplot2 colors will be used. Otherwise, you can specify
a color for each sample. The list can also be named, otherwise, the `id` field
will be used for naming the list.

```{r echo=TRUE, eval=TRUE}
test.path <- "/home/me/recoup_tutorial/chipseq"

chip.input <- list(
    WT_H4K20me1=list(
        id="WT_H4K20me1",
        name="WT H4K20me1",
        file=file.path(test.path,"WT_H4K20me1.bam"),
        format="bam",
        color="#EE0000"
    ),
    Set8KO_H4K20me1=list(
        id="Set8KO_H4K20me1",
        name="Set8KO H4K20me1",
        file=file.path(test.path,"Set8KO_H4K20me1.bam"),
        format="bam",
        color="#00BB00"
    )
)
```

All the following examples will use all your cores if you are using a multicore
system in Linux (not in Windows as the `parallel` package is not supported). If
you want to restrict the number of cores, use the `rc` argument. In real world
examples, you _should_ most of the times restrict the number of cores used as
the memory footprints are quite high in R parallel computations, unless of 
course you have a lot of memory to spare. We are working to improve that at 
least for `recoup`!

### TSS profiles

`recoup` supports several pre-defined and typical locations for creating genomic
profile plots, like transcription start sites (TSSs) and gene bodies. It also 
supports several base-pair resolutions (coverage per base or binned). Have a 
look at the man page for further details.

Firstly, let's point to our custom genomic regions:

```{r echo=TRUE, eval=TRUE}
genome <- file.path(test.path,"mm9_custom_chr12.txt")
```

The following will produce a high resolution (coverage per base-pair) H4K20me1 
WT and Set8 KO TSS profiles. No design is applied. Half of available cores are 
used (`rc` parameter). If the tutorial system has only one core, then the `rc`
parameter is ignored. The flanking region is +/-2kb and no normalization is 
applied (BAM files are sampled from already downsampled initial files).

```{r echo=TRUE, eval=TRUE}
test <- recoup(
    input=chip.input,
    region="tss",
    type="chipseq",
    genome=genome,
    flank=c(2000,2000),
    selector=NULL,
    plotParams=list(plot=FALSE)
)

p <- getr(test,"profile")
plot(p)
h <- getr(test,"heatmap")
draw(h,gap=grid::unit(1,"cm"))
o <- getr(test,"correlation")
plot(o)
```

The following will produce the previous plot at a lower resolution. The output
heatmaps will have a common signal scale. `recoup` objects are resulable so 
that not all fields have to be recalculated (e.g. coverage is calculated once):

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
test2 <- recoup(
    input=test,
    binParams=list(regionBinSize=100),
    plotParams=list(plot=FALSE,heatmapScale="common"),
    saveParams=list(ranges=TRUE,coverage=TRUE,profile=TRUE)
)

p <- getr(test2,"profile")
plot(p)
h <- getr(test2,"heatmap")
draw(h,gap=grid::unit(1,"cm"))
```

The following will produce the previous plot at a medium resolution. The output
heatmaps will have a common signal scale and the brightness will be decreased.
Also, since we know that we have asked for a rather low resolution, we turn the
`forceHeatmapBinning` parameter off and the profile curve also off as we want to
demontrate heatmap brightness.

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
test3 <- recoup(
    input=test,
    region="tss",
    binParams=list(regionBinSize=200,forceHeatmapBinning=FALSE),
    plotParams=list(plot=FALSE,profile=FALSE,heatmapScale="common",
        heatmapFactor=0.5)
)

h <- getr(test2,"heatmap")
draw(h,gap=grid::unit(1,"cm"))
```

Like the previous from a configuration file (not evaluated here):

```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
testc <- recoup(
    input=file.path(test.path,"test_config.txt"),
    region="tss",
    type="chipseq",
    genome=genome,
    flank=c(2000,2000),
    binParams=list(regionBinSize=100),
    selector=NULL
)

#p <- getr(testc,"profile")
#plot(p)
#h <- getr(testc,"heatmap")
#draw(h,gap=grid::unit(1,"cm"))
```

### Gene body profiles

The following will produce a high resolution (in the flanks) H4K20me1 WT and 
Set8 KO gene body profiles. No design is applied. Auto-binning will be applied
as the gene bodies do not have a constant length. Heatmap is not drawn here as
the `forceHeatmapBinning` option will suppress the "unrealistic" (in heatmap
profile only) look of the following:

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
test <- recoup(
    input=chip.input,
    region="genebody",
    type="chipseq",
    genome=genome,
    flank=c(1000,1000),
    binParams=list(flankBinSize=0,regionBinSize=0),
    plotParams=list(plot=FALSE,heatmap=FALSE),
    selector=NULL
)

p <- getr(test,"profile")
plot(p)
```

The previous plot looks a bit unrealistic. To make it look more realistic, we
narrow the flanking regions by lowering the resolution and setting a lower 
number of bins for the flanking regions. Remember that output is reusable:

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
test2 <- recoup(
    input=test,
    binParams=list(flankBinSize=50,regionBinSize=200),
    plotParams=list(plot=FALSE)
)

p <- getr(test2,"profile")
plot(p)
h <- getr(test2,"heatmap")
draw(h,gap=grid::unit(1,"cm"))
```

The following will produce the previous plot with k-means clustering (4 
clusters) applied to the profiles. The reference profile (profile to which
clustering is applied and the resulting order is applied to the other one) is
the WT one. Design is auto-created.

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.height=8}
test3 <- recoup(
    input=test2,
    kmParams=list(k=4,reference="WT_H4K20me1")
)

p <- getr(test3,"profile")
plot(p)
h <- getr(test3,"heatmap")
draw(h,gap=grid::unit(1,"cm"))
o <- getr(test3,"profile")
plot(o)
```

Let's now look at plots with designs. The following will produce plots like the
first ones but with a design according to gene expression levels and strand 
applied to facet the profiles:

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.height=8}
design <- file.path(test.path,"design.txt")

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
test4 <- recoup(
    input=test2,
    design=design
)

p <- getr(test4,"profile")
plot(p)
h <- getr(test4,"heatmap")
draw(h,gap=grid::unit(1,"cm"))
o <- getr(test3,"profile")
plot(o)
```

With recoup, you can also apply several orderings of the profiles, like from
the maximum coverage to the minimum coverage profile, hierarchical clustering
and others. For further details have a look at the `recoup` man page. The 
following will order the profiles according to the sum of the coverage of the
WT profile in decreasing order. You cal also try the commented out options:

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
test5 <- recoup(
    input=chip.input,
    region="tss",
    type="chipseq",
    genome=genome,
    flank=c(2000,2000),
    orderBy=list(what="sum1"),
    #orderBy=list(what="hc1"),
    #orderBy=list(what="sum2",order="descending"),
    #orderBy=list(what="max2",order="ascending"),
    #orderBy=list(what="suma",order="descending"),
    binParams=list(flankBinSize=50,regionBinSize=100,sumStat="mean",
        smooth=TRUE,forceHeatmapBinning=FALSE),
    plotParams=list(plot=FALSE),
    selector=NULL
)

h <- getr(test5,"heatmap")
draw(h,gap=grid::unit(1,"cm"))
```

Like above but with a design (reduced by one factor to reduce heatmap plot
complexity) and hierarchical clustering ordering (1st profile is reference):

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
design1 <- read.delim(file.path(test.path,"design.txt"),row.names=1)
d1 <- as.data.frame(design1[,2])
rownames(d1) <- rownames(design1)

test6 <- recoup(
    input=test5,
    design=d1,
    orderBy=list(what="hc1"),
    binParams=list(flankBinSize=50,regionBinSize=100,sumStat="mean",
        smooth=TRUE,forceHeatmapBinning=FALSE),
    plotParams=list(plot=FALSE,profile=FALSE)
)

h <- getr(test6,"heatmap")
draw(h,gap=grid::unit(1,"cm"))
```
__Note 1__: `recoup` list objects can be subsetted/sliced. Have a look at the
man page of the `sliceObj` function.

__Note 2__: In all cases, the ordering of the profiles as well as well as the 
elements (genomic areas) belonging to a group wheh using design files for 
grouping, can be acquired either from the `rownames` of the `profile` member
of the `recoup$data` output list object, or from the `design` member of the 
`recoup` output object. From the `design` member you can also extract the number
of elements in each group, which is very useful especially in the case of 
applying k-means clustering to the profiles.

## R session information

```{r echo=FALSE}
sessionInfo()
```
